### ISSUE TITLE
HTTP 500 originating from delegate script

### ISSUE THREAD

**Issue by adolski** [medusa-project/cantaloupe#114]
I have been able to get traces for two distinct HTTP 500s originating from the delegate script mechanism. They are not easy for me to reproduce, as they only happen rarely amid large numbers of concurrent requests.

(This is on the `develop` branch, but `release/3.3` might be affected too.)

Here is a little Node script for testing:

```javascript
IMAGE = 'test.jpg';

var http = require('http');

for (var i = 0; i < 100; i++) {
  var request = http.request({
      host: 'localhost',
      port: 8182,
      method: 'GET',
      path: '/iiif/2/' + IMAGE + '/full/full/' + i + '/default.jpg'
  });

  request.end();

  request.on('response', function(response) {
    if (response.statusCode != 200) {
      console.log(response.statusCode);
      console.log(JSON.stringify(response.headers));
    }
  });
}
```

Here is a trace involving the `overlay()` delegate method. (My implementation only returns a hard-coded hash.)

```
12:14:43.450 [qtp2034182655-83] ERROR e.i.l.c.r.AbstractResource - java.lang.ClassCastException: org.jruby.java.proxies.ConcreteJavaProxy cannot be cast to org.jruby.RubyArray
javax.script.ScriptException: java.lang.ClassCastException: org.jruby.java.proxies.ConcreteJavaProxy cannot be cast to org.jruby.RubyArray
	at org.jruby.embed.jsr223.JRubyEngine.wrapException(JRubyEngine.java:104)
	at org.jruby.embed.jsr223.JRubyEngine.eval(JRubyEngine.java:93)
	at org.jruby.embed.jsr223.JRubyEngine.eval(JRubyEngine.java:142)
	at edu.illinois.library.cantaloupe.script.RubyScriptEngine.doInvoke(RubyScriptEngine.java:165)
	at edu.illinois.library.cantaloupe.script.RubyScriptEngine.invoke(RubyScriptEngine.java:125)
	at edu.illinois.library.cantaloupe.operation.overlay.DelegateOverlayService.overlayProperties(DelegateOverlayService.java:150)
	at edu.illinois.library.cantaloupe.operation.overlay.DelegateOverlayService.getOverlay(DelegateOverlayService.java:43)
	at edu.illinois.library.cantaloupe.operation.overlay.OverlayService.newOverlay(OverlayService.java:79)
	at edu.illinois.library.cantaloupe.resource.AbstractResource.addNonEndpointOperations(AbstractResource.java:311)
	at edu.illinois.library.cantaloupe.resource.iiif.v2.ImageResource.doGet(ImageResource.java:145)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.restlet.resource.ServerResource.doHandle(ServerResource.java:523)
	at org.restlet.resource.ServerResource.get(ServerResource.java:742)
	at org.restlet.resource.ServerResource.doHandle(ServerResource.java:617)
	at org.restlet.resource.ServerResource.doNegotiatedHandle(ServerResource.java:678)
	at org.restlet.resource.ServerResource.doConditionalHandle(ServerResource.java:356)
	at org.restlet.resource.ServerResource.handle(ServerResource.java:1043)
	at org.restlet.resource.Finder.handle(Finder.java:236)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Router.doHandle(Router.java:422)
	at org.restlet.routing.Router.handle(Router.java:641)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.engine.application.StatusFilter.doHandle(StatusFilter.java:140)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.engine.CompositeHelper.handle(CompositeHelper.java:202)
	at org.restlet.engine.application.ApplicationHelper.handle(ApplicationHelper.java:77)
	at org.restlet.Application.handle(Application.java:385)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Router.doHandle(Router.java:422)
	at org.restlet.routing.Router.handle(Router.java:641)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Router.doHandle(Router.java:422)
	at org.restlet.routing.Router.handle(Router.java:641)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.engine.CompositeHelper.handle(CompositeHelper.java:202)
	at org.restlet.Component.handle(Component.java:408)
	at org.restlet.Server.handle(Server.java:507)
	at org.restlet.engine.connector.ServerHelper.handle(ServerHelper.java:63)
	at org.restlet.engine.adapter.HttpServerHelper.handle(HttpServerHelper.java:143)
	at org.restlet.ext.servlet.ServerServlet.service(ServerServlet.java:1117)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:840)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:584)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1180)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at org.eclipse.jetty.server.Server.handle(Server.java:534)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:320)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:283)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:108)
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassCastException: org.jruby.java.proxies.ConcreteJavaProxy cannot be cast to org.jruby.RubyArray
	at org.jruby.embed.variable.Argv.getJavaObject(Argv.java:197)
	at org.jruby.embed.internal.BiVariableMap.put(BiVariableMap.java:337)
	at org.jruby.embed.ScriptingContainer.put(ScriptingContainer.java:1185)
	at org.jruby.embed.jsr223.Utils.put(Utils.java:234)
	at org.jruby.embed.jsr223.Utils.preEval(Utils.java:109)
	at org.jruby.embed.jsr223.JRubyEngine.eval(JRubyEngine.java:88)
	... 79 common frames omitted
```

And the one below happened only once out of 10s of thousands of requests:

```
ERROR o.r.C.WebApplication - Exception or error caught in server resource
org.restlet.resource.ResourceException: Internal Server Error (500) - The server encountered an unexpected condition which prevented it from fulfilling the request
	at org.restlet.resource.ServerResource.doHandle(ServerResource.java:539)
	at org.restlet.resource.ServerResource.get(ServerResource.java:742)
	at org.restlet.resource.ServerResource.doHandle(ServerResource.java:617)
	at org.restlet.resource.ServerResource.doNegotiatedHandle(ServerResource.java:678)
	at org.restlet.resource.ServerResource.doConditionalHandle(ServerResource.java:356)
	at org.restlet.resource.ServerResource.handle(ServerResource.java:1043)
	at org.restlet.resource.Finder.handle(Finder.java:236)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Router.doHandle(Router.java:422)
	at org.restlet.routing.Router.handle(Router.java:641)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.engine.application.StatusFilter.doHandle(StatusFilter.java:140)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.engine.CompositeHelper.handle(CompositeHelper.java:202)
	at org.restlet.engine.application.ApplicationHelper.handle(ApplicationHelper.java:77)
	at org.restlet.Application.handle(Application.java:385)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Router.doHandle(Router.java:422)
	at org.restlet.routing.Router.handle(Router.java:641)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.routing.Router.doHandle(Router.java:422)
	at org.restlet.routing.Router.handle(Router.java:641)
	at org.restlet.routing.Filter.doHandle(Filter.java:150)
	at org.restlet.routing.Filter.handle(Filter.java:197)
	at org.restlet.engine.CompositeHelper.handle(CompositeHelper.java:202)
	at org.restlet.Component.handle(Component.java:408)
	at org.restlet.Server.handle(Server.java:507)
	at org.restlet.engine.connector.ServerHelper.handle(ServerHelper.java:63)
	at org.restlet.engine.adapter.HttpServerHelper.handle(HttpServerHelper.java:143)
	at org.restlet.ext.servlet.ServerServlet.service(ServerServlet.java:1117)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:840)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:584)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1180)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at org.eclipse.jetty.server.Server.handle(Server.java:534)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:320)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:283)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:108)
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException: null
	at org.jruby.embed.jsr223.JRubyEngine.invokeMethod(JRubyEngine.java:217)
	at edu.illinois.library.cantaloupe.script.RubyScriptEngine.doInvoke(RubyScriptEngine.java:164)
	at edu.illinois.library.cantaloupe.script.RubyScriptEngine.invoke(RubyScriptEngine.java:125)
	at edu.illinois.library.cantaloupe.resource.AbstractResource.checkAuthorization(AbstractResource.java:402)
	at edu.illinois.library.cantaloupe.resource.iiif.v2.ImageResource.doGet(ImageResource.java:115)
	at sun.reflect.GeneratedMethodAccessor63.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.restlet.resource.ServerResource.doHandle(ServerResource.java:523)
	... 66 common frames omitted
```

**Comment by devanonyme** [comment]
It does look like the error I was observing (see the sample below) but I was not able to figure out where to put breakpoints or try/catch to pinpoint thoses 

500 Internal Server Error
java.lang.ClassCastException: org.jruby.java.proxies.ConcreteJavaProxy cannot be cast to org.jruby.RubyArray

javax.script.ScriptException: java.lang.ClassCastException: org.jruby.java.proxies.ConcreteJavaProxy cannot be cast to org.jruby.RubyArray
	at org.jruby.embed.jsr223.JRubyEngine.wrapException(JRubyEngine.java:104)
	at org.jruby.embed.jsr223.JRubyEngine.eval(JRubyEngine.java:93)
	at org.jruby.embed.jsr223.JRubyEngine.eval(JRubyEngine.java:142)

**Comment by devanonyme** [comment]
With my manual html test I'm able to reproduce in more or less 10 reload. I can maybe give a try with the node test and see what I get...

**Comment by devanonyme** [comment]
I was not able to reproduce it so far with the node.js test . I hacked it a bit it to add similar header and tried different concurrency / loop setup (Chrome is supposed to do a limited number of concurrent request by ex. so my guess it that the trouble is not related to a very high number of concurrent images (my html test have 20 of them). Still able to reproduce it relatively easily with the Browser (Chrome, Explorer, it seem to be harder with Firefox but I don't think the issue is browser specific).

**Comment by devanonyme** [comment]
I was able to find a place to place a breakpoint that seem to be related to the issue. I will try to understand what could happening from here, I'm no sure I will be able to improve reproducibility

![reproduce_close](https://cloud.githubusercontent.com/assets/21067311/26562416/0a4b32e8-4494-11e7-842a-9b48eb9b4101.PNG)

**Comment by adolski** [comment]
I'm almost certain this was fixed by 73d16901f67c2c716681cc8d317a518d503af776, which was merged into 3.3.1. I can make it happen in 3.3, but not 3.3.1. Please reopen if you ever see it crop up again.

**Comment by devanonyme** [comment]
I think we still have issues but my testing was pushed by other priorities. I actually improved my solution on the application side (limit concurrent call + reload attempt), so it's not high priority for now.

I'll reopen as needed, no problem!
